tests/ui/unnecessary_indexing.rs

```diff
@@ -0,0 +1,131 @@
+//@no-rustfix
+#![allow(unused)]
+#![allow(dropping_copy_types)]
+#![allow(dropping_references)]
+#![warn(clippy::unnecessary_indexing)]
+
+fn c(x: i32) -> i32 {
+    println!("{x}");
+    10
+}
+
+struct Struct;
+impl Struct {
+    pub fn a(x: i32) -> i32 {
+        println!("{x}");
+        10
+    }
+}
+
+fn main() {
+    // lint on vecs with a call
+    let a: Vec<i32> = vec![1];
+    if !a.is_empty() {
+        let b = c(a[0]);
+    }
+
+    // lint on vecs with a method call
+    let a: Vec<i32> = vec![1];
+    if !a.is_empty() {
+        let b = Struct::a(a[0]);
+    }
+
+    // lint on arrays with a call
+    let a: &[i32] = &[1];
+    if !a.is_empty() {
+        let b = c(a[0]);
+    }
+
+    // lint on arrays with a method call
+    let a: &[i32] = &[1];
+    if !a.is_empty() {
+        let b = Struct::a(a[0]);
+    }
+
+    // lint on vecs with a local access
+    let a: Vec<i32> = vec![1];
+    if !a.is_empty() {
+        let b = a[0];
+    }
+
+    // lint on arrays with a local access
+    let a: &[i32] = &[1];
+    if !a.is_empty() {
+        let b = a[0];
+    }
+
+    // lint when access is not first line
+    let a: &[i32] = &[1];
+    if !a.is_empty() {
+        dbg!(a);
+        let b = a[0];
+    }
+
+    // lint on multiple accesses/locals
+    let a: &[i32] = &[1];
+    if !a.is_empty() {
+        dbg!(a);
+        let b = a[0];
+        let c = a[0];
+        drop(a[0]);
+    }
+
+    // lint on multiple accesses
+    let a: &[i32] = &[1];
+    if !a.is_empty() {
+        dbg!(a);
+        drop(a[0]);
+        drop(a[0]);
+    }
+
+    // dont lint when not accessing [0]
+    let a: &[i32] = &[1, 2];
+    if !a.is_empty() {
+        let b = a[1];
+    }
+
+    // dont lint when access is dynamic
+    const T: usize = 0;
+
+    let a: &[i32] = &[1];
+    if !a.is_empty() {
+        let b = a[T];
+    }
+
+    // dont lint without unary
+    let a: &[i32] = &[1];
+    if a.is_empty() {
+        let b = a[0];
+    }
+
+    // dont lint if we have mutable reference
+    let mut a: &[i32] = &[1];
+    if !a.is_empty() {
+        drop(&mut a);
+        let b = a[0];
+    }
+
+    // dont lint if we have a mutable reference, even if the mutable reference occurs after what we are
+    // linting against
+    let mut a: &[i32] = &[1];
+    if !a.is_empty() {
+        let b = a[0];
+        drop(&mut a);
+    }
+
+    // dont lint on mutable auto borrow
+    let mut a = vec![1, 2, 3];
+    if !a.is_empty() {
+        a.push(1);
+        let b = a[0];
+        b;
+    }
+
+    // do not lint if conditional receiver is mutable reference
+    let a = &mut vec![1, 2, 3];
+    if !a.is_empty() {
+        let b = a[0];
+        a;
+        b;
+    }
```

Comment:
It occurs to me now (sorry 😅) that this issue & some others actually came from fact that we're changing the type of `b` from `T` to `&T` and creating a borrow that didn't exist before. The suggestions do include the caveat that dereferences may be required but it would be good to make them correct if we can

A possible route would be to split `extra_exprs` into two variables, one for indexes that are taken by reference (parent is a ref `&x[0]` or by auto borrow `x[0].len()`) and one for indexes that are copied

If all usages are copied then we can suggest `if let Some(&b) = a.first()`

Reply:
I just added this based on what you suggested, it's hard to know if this will really catch everything or not or if it's susceptible to FPs but I imagine they are rare. All the tests I wrote beforehand seem to align with these changes properly and I do agree it's nice to provide correct suggestions where possible.